File: clients/typescript/ctxd-client/src/events.ts
/**
 * CloudEvents-shaped `Event` type + canonical-bytes helper.
 *
 * The shape mirrors `crates/ctxd-core/src/event.rs` and the JSON
 * Schema at `docs/api/events.schema.json`. We keep IDs as plain
 * strings (UUIDv7 lexical form) rather than parsing into a UUID type;
 * coercing risks normalizing to canonical lowercase and drifting from
 * the daemon's wire encoding.
 *
 * The wire form omits empty optional fields (matches the Rust
 * `skip_serializing_if`). The canonical signing form is stricter:
 * always includes `parents` (sorted, possibly `[]`), always includes
 * `attestation` (possibly `null`), and excludes `predecessorhash` +
 * `signature`.
 */

/**
 * A CloudEvents v1.0 event with ctxd extensions.
 *
 * `type` is the wire field name (CloudEvents discriminator). We expose
 * it under that name to match the on-the-wire JSON exactly — callers
 * who want a TypeScript-friendly alias can destructure as
 * `{ type: eventType, ...rest }`.
 */
export interface Event {
  /** CloudEvents spec version. Always `"1.0"` for v0.3. */
  specversion: string;
  /** Globally-unique UUIDv7 string. */
  id: string;
  /** Identifies the context in which the event happened. */
  source: string;
  /** Subject path the event is filed under. */
  subject: string;
  /** Event type discriminator (e.g. `ctx.note`). */
  type: string;
  /** RFC3339 timestamp of when the event was created. */
  time: string;
  /** Content type of `data`. */
  datacontenttype: string;
  /** Event payload. JSON of any shape. */
  data: unknown;
  /** SHA-256 hash of the predecessor event's canonical form. */
  predecessorhash?: string | null;
  /** Ed25519 signature over the canonical form, hex-encoded. */
  signature?: string | null;
  /** Parent event ids (UUIDv7 strings). */
  parents?: string[];
  /** Optional TEE attestation blob, hex-encoded. */
  attestation?: string | null;
}

/**
 * Produce the wire-shape JSON object for `event` (omitting empty
 * optional fields).
 *
 * - `predecessorhash`, `signature`, `attestation` are omitted when
 *   `null` / `undefined`.
 * - `parents` is omitted when empty (matches the Rust
 *   `skip_serializing_if = "Vec::is_empty"`).
 */
export function eventToWire(event: Event): Record<string, unknown> {
  const out: Record<string, unknown> = {
    specversion: event.specversion,
    id: event.id,
    source: event.source,
    subject: event.subject,
    type: event.type,
    time: event.time,
    datacontenttype: event.datacontenttype,
    data: event.data,
  };
  if (event.predecessorhash != null) {
    out.predecessorhash = event.predecessorhash;
  }
  if (event.signature != null) {
    out.signature = event.signature;
  }
  if (event.parents && event.parents.length > 0) {
    out.parents = [...event.parents];
  }
  if (event.attestation != null) {
    out.attestation = event.attestation;
  }
  return out;
}

/**
 * Parse an unknown value into an `Event`. Throws on missing required
 * fields. We deliberately avoid a heavyweight schema validator (zod,
 * io-ts) to keep the dependency surface tight — the wire shape is
 * narrow and stable, and a hand-written guard is easier to audit.
 */
export function parseEvent(raw: unknown): Event {
  if (raw == null || typeof raw !== "object") {
    throw new TypeError(`event must be an object, got ${typeof raw}`);
  }
  const r = raw as Record<string, unknown>;
  const required = ["id", "source", "subject", "type", "time", "data"] as const;
  for (const key of required) {
    if (!(key in r)) {
      throw new TypeError(`event missing required field: ${key}`);
    }
  }
  const ev: Event = {
    specversion: typeof r.specversion === "string" ? r.specversion : "1.0",
    id: String(r.id),
    source: String(r.source),
    subject: String(r.subject),
    type: String(r.type),
    time: String(r.time),
    datacontenttype:
      typeof r.datacontenttype === "string"
        ? r.datacontenttype
        : "application/json",
    data: r.data,
  };
  if (typeof r.predecessorhash === "string") {
    ev.predecessorhash = r.predecessorhash;
  }
  if (typeof r.signature === "string") {
    ev.signature = r.signature;
  }
  if (Array.isArray(r.parents)) {
    ev.parents = r.parents.map(String);
  }
  if (typeof r.attestation === "string") {
    ev.attestation = r.attestation;
  }
  return ev;
}

/**
 * Produce the canonical signing bytes for `event`.
 *
 * Mirrors `ctxd_core::signing::canonical_bytes` exactly: a JSON
 * object with **sorted keys** containing every CloudEvents field
 * except `predecessorhash` and `signature`, plus `parents` (sorted
 * lexicographically; empty array if none) and `attestation`
 * (hex-encoded or `null`).
 *
 * The Rust side emits this via `serde_json::to_vec(BTreeMap<&str,
 * Value>)`: compact JSON (no spaces) with sorted keys. We construct
 * the same shape and rely on the helper below for sort-order
 * stability — `JSON.stringify` does NOT sort keys by default.
 */
export function canonicalBytes(event: Event): Uint8Array {
  const parents = [...(event.parents ?? [])].sort();
  const attestation = event.attestation ?? null;
  const payload: Record<string, unknown> = {
    attestation,
    data: event.data,
    datacontenttype: event.datacontenttype,
    id: event.id,
    parents,
    source: event.source,
    specversion: event.specversion,
    subject: event.subject,
    time: event.time,
    type: event.type,
  };
  const json = stableStringify(payload);
  return new TextEncoder().encode(json);
}

/**
 * Stable, sorted-key JSON serialization (compact form).
 *
 * Mirrors `serde_json` over `BTreeMap<&str, Value>` byte-for-byte:
 * sorted keys, no whitespace, no trailing newline. Pure recursive
 * walk; Maps and Sets are not handled (the SDK never emits them in a
 * canonical form).
 */
function stableStringify(value: unknown): string {
  if (value === null || value === undefined) return "null";
  if (typeof value === "number") {
    // Match serde_json: integers without trailing decimal, finite only.
    if (!Number.isFinite(value)) {
      throw new TypeError("non-finite number in canonical form");
    }
    return JSON.stringify(value);
  }
  if (typeof value === "string" || typeof value === "boolean") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return "[" + value.map(stableStringify).join(",") + "]";
  }
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).sort();
    const parts: string[] = [];
    for (const k of keys) {
      parts.push(JSON.stringify(k) + ":" + stableStringify(obj[k]));
    }
    return "{" + parts.join(",") + "}";
  }
  // bigint, symbol, function — not supported in canonical JSON.
  throw new TypeError(`unsupported value in canonical form: ${typeof value}`);
}

File: clients/python/ctxd-py/src/ctxd/_events.py
"""CloudEvents-shaped :class:`Event` model + canonical-bytes helper.

The :class:`Event` shape mirrors ``crates/ctxd-core/src/event.rs`` and
the JSON Schema at ``docs/api/events.schema.json``. We keep IDs as
plain strings (UUIDv7 lexical form) to guarantee byte-exact round-trip
through pydantic — coercing to ``uuid.UUID`` would normalize to
canonical lowercase and risk drifting from the wire encoding the
daemon emits.

Optional fields use ``model_config["json_schema_extra"]`` semantics: a
field that is ``None`` / empty list is omitted on the wire, but always
serialized in the canonical form used for hashing and signing.
"""

from __future__ import annotations

import json
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_serializer


class Event(BaseModel):
    """A CloudEvents v1.0 event with ctxd extensions.

    ``event_type`` is the Python field name; on the wire it serializes
    as ``type`` per the CloudEvents spec. Optional fields are omitted
    from the JSON output when absent / empty.
    """

    model_config = ConfigDict(
        # Allow ``type`` to populate ``event_type`` on parse, and the
        # serializer alias to emit ``type`` on dump.
        populate_by_name=True,
        # We use ``additionalProperties: false`` semantics on the wire,
        # but accept unknown fields silently to forward-compat with new
        # daemon releases.
        extra="ignore",
    )

    specversion: str = "1.0"
    """CloudEvents spec version. Always ``"1.0"`` for v0.3."""

    id: str
    """Globally-unique UUIDv7 string."""

    source: str
    """Identifies the context in which the event happened."""

    subject: str
    """Subject path the event is filed under."""

    event_type: str = Field(alias="type")
    """Event type discriminator (e.g. ``ctx.note``)."""

    time: datetime
    """RFC3339 timestamp of when the event was created."""

    datacontenttype: str = "application/json"
    """Content type of ``data``."""

    data: Any
    """Event payload. JSON of any shape."""

    predecessorhash: str | None = None
    """SHA-256 hash of the predecessor event's canonical form."""

    signature: str | None = None
    """Ed25519 signature over the canonical form, hex-encoded."""

    parents: list[str] = Field(default_factory=list)
    """Parent event ids (UUIDv7 strings)."""

    attestation: str | None = None
    """Optional TEE attestation blob, hex-encoded."""

    @field_serializer("time")
    def _serialize_time(self, t: datetime) -> str:
        """Serialize time as RFC3339 with ``Z`` suffix when UTC.

        The Rust daemon emits ``2026-01-01T00:00:00Z`` (no offset
        suffix) for UTC times. We match that shape so a round-trip of a
        daemon-produced event is byte-stable.
        """
        return _format_rfc3339(t)

    def model_dump_wire(self) -> dict[str, Any]:
        """Dump the event in wire form (omitting empty optional fields).

        - ``predecessorhash``, ``signature``, ``attestation`` are
          omitted when ``None``.
        - ``parents`` is omitted when empty (matches the Rust
          ``skip_serializing_if = "Vec::is_empty"``).
        """
        out: dict[str, Any] = {
            "specversion": self.specversion,
            "id": self.id,
            "source": self.source,
            "subject": self.subject,
            "type": self.event_type,
            "time": _format_rfc3339(self.time),
            "datacontenttype": self.datacontenttype,
            "data": self.data,
        }
        if self.predecessorhash is not None:
            out["predecessorhash"] = self.predecessorhash
        if self.signature is not None:
            out["signature"] = self.signature
        if self.parents:
            out["parents"] = list(self.parents)
        if self.attestation is not None:
            out["attestation"] = self.attestation
        return out


def _format_rfc3339(t: datetime) -> str:
    """Emit ``YYYY-MM-DDTHH:MM:SS[.fff]Z`` for UTC, keep offset for non-UTC.

    The daemon serializes ``DateTime<Utc>`` via chrono, which emits
    ``Z`` for UTC and the offset for other timezones. Python's
    ``datetime.isoformat`` emits ``+00:00`` for UTC by default — we
    convert that to ``Z`` so round-trips are byte-stable.
    """
    if t.tzinfo is None:
        # Treat naive datetimes as UTC. The daemon never emits naive
        # times; this branch only protects callers that hand-build
        # Events without a tzinfo.
        s = t.isoformat() + "Z"
        return s
    iso = t.isoformat()
    if iso.endswith("+00:00"):
        return iso[:-6] + "Z"
    return iso


def canonical_bytes(event: Event) -> bytes:
    """Produce the canonical signing bytes for ``event``.

    Mirrors ``ctxd_core::signing::canonical_bytes`` exactly: a JSON
    object with **sorted keys** containing every CloudEvents field
    except ``predecessorhash`` and ``signature``, plus the v0.3
    ``parents`` (sorted lexicographically by string id; empty array if
    none) and ``attestation`` (hex-encoded or ``null``) fields.

    The output is `serde_json::to_vec(BTreeMap<&str, Value>)` on the
    Rust side: compact JSON (no spaces) with sorted keys.
    """
    parents_sorted = sorted(event.parents)
    attestation_val: str | None = event.attestation  # already hex string
    payload = {
        "attestation": attestation_val,
        "data": event.data,
        "datacontenttype": event.datacontenttype,
        "id": event.id,
        "parents": parents_sorted,
        "source": event.source,
        "specversion": event.specversion,
        "subject": event.subject,
        "time": _format_rfc3339(event.time),
        "type": event.event_type,
    }
    # ``sort_keys=True`` matches BTreeMap iteration order on the Rust
    # side. ``separators=(",", ":")`` matches serde_json's compact
    # output (no whitespace).
    return json.dumps(payload, sort_keys=True, separators=(",", ":")).encode("utf-8")
